README.md
# Norsu

A web extension blockchain wallet for the TAC platform.

## Build instructions

Run `./build.sh`, which installs the required node modules, does an in-place edit on one of the npm packages to change the chain Id from 1 to 15, and then builds the distribution packages. These can be found in the `dist` folder.

## Installing the extension

Until Google approves the extension, it has to be installed manually.

Open Chrome and enter `chrome://extensions` in the address bar (for Brave, change that to `brave://extensions`).

Toggle the "Developer mode" switch in the top-right corner to on.

Click on "Load unpacked" and select the `./dist/chrome` folder.

The extension should now be loaded into your browser.

## Web3 instructions

The following examples explain how to produce Web3 pages for websites that interact with the wallet. I haven't tested these yet, and the object names need to be changed, but for now you can play around with the wallet using the text below as a guide:

On browser pages that operate with https or http, with the Norsu extension installed, Norsu global object becomes available, featuring the following methods:

- `auth`
- `publicState`
- `signAndPublishCancelOrder`
- `signAndPublishOrder`
- `signAndPublishTransaction`
- `signCancelOrder`
- `signOrder`
- `signTransaction`
- `signRequest`
- `signTransactionPackage`
- `signCustomData`
- `verifyCustomData`
- `notification`
- `encryptMessage`
- `decryptMessage`
- `resourceIsApproved`
- `resourceIsBlocked`
- `on`

All methods, except for "on" operate asynchronously and return [promises](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Promise).

In code, you can use [TypeScript types](https://github.com/tacplatform/tackeeper-types)

On initialize `window.Norsu` has not api methods.
You can use `Norsu.initialPromise` for waiting end initializing api.
Example:

```js
Norsu.initialPromise.then(keeperApi => {
  /*...init app*/
  keeperApi.publicState().then(state => startApp(state));
});
```

In Norsu, for greater security and ease of use, each new website using API has to be allowed by the user. At the first attempt to use API (except "`on"`), the user will see a request to allow that website's access to Norsu. If the user agrees to allow access, the website is considered trusted and can use API on its pages. Otherwise, the website is blocked, and an error message will be displayed in response to all requests `{message: "Api rejected by user", code: 12}`. The users won't see new notifications. To grant access, the user will mark the website as trusted in the interface.

### Description of methods

#### publicState

If a website is trusted, Norsu public data are returned.

Example:

```js
Norsu.publicState()
  .then(state => {
    console.log(state); //displaying the result in the console
    /*...processing data */
  })
  .catch(error => {
    console.error(error); // displaying the result in the console
    /*...processing errors */
  });
```

or

```js
const getPublicState = async () => {
  try {
    const state = await Norsu.publicState();
    console.log(state); // displaying the result in the console
    /*... processing data*/
  } catch (error) {
    console.error(error); // displaying the result in the console
    /*... processing errors */
  }
};

const result = await getPublicState();
```

REPLY

```json
{
    "initialized": true,
    "locked": true,
    "account": {
        "name": "foo",
        "publicKey": "bar",
        "address": "tac address",
        "networkCode": "network byte",
        "balance": {
            "available": "balance in tac",
            "leasedOut": "leased balance"
        }
    },
    "network": {
        "code": "W",
        "server": "https://nodes.tacnodes.com/",
        "matcher": "https://matcher.tac.exchange/"
    },
    "messages": [],
    "txVersion": {
        "3": [ 2 ],
        "4": [ 2 ],
        "5": [ 2 ],
        ...
    }
}
```

Description of query return fields

- `initialized` - boolean keeper initialized
- `locked` - boolean keeper in wait mode
- `account` – current account, if the user allowed access to the website, or null
- `network` – current Tac network, node and matcher addresses
- `messages` – signature request statuses
- `txVersion` – available transaction versions for each type

Possible errors

- `{ message: "Init Norsu and add account" }` – Norsu is not initialized
- `{ message: "Add Norsu account" }` – Norsu accessed, but there are no accounts
- `{ message: "User denied message" }` – the user denied the website operation with Norsu

#### notification

Send message to keeper.
Ypu can send message only 1 time in 30 sec for trusted sites with send permission.

`notification` facilitates input of the following data

- `title` - string (20 chars max) (required field)
- `message` - string (250 chars max) (optional field)

return Promise

Example:

```js
Norsu.notification({
  title: 'Hello!',
  message: 'Congratulation!!!',
});
```

Possible errors

- `{message: "Incorrect notification data", data: "title has more than 20 characters", code: "19"}` - Incorrect notification title
- `{message: "Incorrect notification data", data: null, code: "19"}` - Incorrect notification data
- `{message: "Can't sent notification", data: {msg: "Min notification interval 30s. Wait 28.017s."}, code: "18"}` - try to send later, you can send 1 message in 30 sec
- `{message: "Api rejected by user", code: 12}` the user denied the request or the website is not trusted.

#### encryptMessage

You can encrypt string messages to account in Tac network.
You need have recipient publicKey.

Norsu.encryptMessage(`*string to encrypt*`, `*public key in base58 string*`, `*prefix is secret app string need for encoding*`)

Example:

```js
Norsu.encryptMessage(
  'My message',
  '416z9d8DQDy5MPTqDhvReRBaPb19gEyVRWvHcewpP6Nc',
  'my app'
).then(encryptedMessage => {
  console.log(encryptedMessage);
});
```

Possible errors

- `{ message: "Init Norsu and add account" }` – Norsu is not initialized
- `{ message: "App is locked" }` – Norsu is locked
- `{ message: "Add Norsu account" }` – Norsu accessed, but there are no accounts
- `{ message: "User denied message" }` – the user denied the website operation with Norsu

#### decryptMessage

You can decrypt string messages from account in Tac network to you.
You need have sender publicKey and encrypted message.

Norsu.decryptMessage(`*string to decrypt*`, `*public key in base58 string*`, `*prefix is secret app string need for encoding*`)

Example:

```js
Norsu.decryptMessage(
  '**encrypted msg**',
  '416z9d8DQDy5MPTqDhvReRBaPb19gEyVRWvHcewpP6Nc',
  'my app'
).then(message => {
  console.log(message);
});
```

Possible errors

- `{ message: "Init Norsu and add account" }` – Norsu is not initialized
- `{ message: "App is locked" }` – Norsu is locked
- `{ message: "Add Norsu account" }` – Norsu accessed, but there are no accounts
- `{ message: "User denied message" }` – the user denied the website operation with Norsu

#### on

Allows subscribing to Norsu events.

Supports events:

- `update` – subscribe for updates of the state

Example:

```js
Norsu.on('update', state => {
  //state object as from Norsu.publicState
});
```

If a website is not trusted, events won't show.

#### auth

This is a method for obtaining a signature of authorization data while verifying Tac' user. It works the same way as [Web Auth API](https://docs.tac.exchange/en/tac-exchange/tac-exchange-client-api/tac-exchange-web-auth-api).

Example:

```js
const authData = { data: 'Auth on my site' };
Norsu.auth(authData)
  .then(auth => {
    console.log(auth); //displaying the result on the console
    /*...processing data */
  })
  .catch(error => {
    console.error(error); // displaying the result on the console
    /*...processing errors */
  });
```

or

```js
const getAuthData = async authData => {
  try {
    const state = await Norsu.auth(authData);
    console.log(state); // displaying the result on the console
    /*... processing data */
  } catch (error) {
    console.error(error); // displaying the result on the console
    /*... processing errors */
  }
};

const authData = { data: 'Auth on my site' };
getAuthData(authData);
```

`auth` facilitates input of the following data

- `name` – name of the service (optional field)
- `data` – a string line with any data (required field)
- `referrer` – a websites' full URL for redirect (optional field)
- `icon` – path to the logo relative to the `referrer`or origin of the website (optional field)
- `successPath` – relative path to the website's Auth API (optional field)

Example

```js
const authData = {
  data: 'Generated string from server',
  name: 'My test App',
  icon: '/img/icons/tac_logo.svg',
  referrer: 'https://tac.exchange/',
  successPath: 'login',
};

Norsu.auth(authData)
  .then(data => {
    //data – data from Norsu
    //verifying signature and saving the address...
    console.log(data);
  })
  .catch(error => {
    //processing the error
  });
```

If the verification is successful, Norsu will return to the promise an object containing data for signature verification:

- `host` – a host that requested a signature
- `name` – the name of an application that requested a signature
- `prefix` – a prefix participating in the signature
- `address` – an address in Tac network
- `publicKey` – the user's public key
- `signature` - signature
- `version` – API version

ERRORS

- `{message: "Invalid data", data: "[{"field":"data","type":"string","message":"field is required"}]", code: 9}` – signature data contain errors
- `{message: "User denied message", code: 10}` – the user denied the request
- `{message: "Api rejected by user", code: 12}` - the website is not trusted

#### signTransaction

A method for signing transactions in Tac' network.

Example:

```js
const txData = {
  type: 4,
  data: {
    amount: {
      assetId: 'TAC',
      tokens: '1.567',
    },
    fee: {
      assetId: 'TAC',
      tokens: '0.001',
    },
    recipient: 'test',
  },
};
Norsu.signTransaction(txData)
  .then(data => {
    //data – a line ready for sending to Tac network's node (server)
  })
  .catch(error => {
    //Processing errors
  });
```

API returns lines, not an object, as in javascript precision is lost in operation with 8-byte integers.

A description of supported transaction types is below.

In the example, we are signing a transaction for transferring TAC to the alias `test`in Tac' network.

REPLY

```json
{
  "version": 2,
  "assetId": "",
  "amount": 156700000,
  "feeAssetId": "",
  "fee": 100000,
  "recipient": "recipient",
  "attachment": "",
  "timestamp": 1548770230589,
  "senderPublicKey": "public key",
  "proofs": ["signature"],
  "type": 4
}
```

ERRORS

- `{message: "User denied message", code: 10}` – The user denied the request.
- `{message: "Api rejected by user", code: 12}` – The website is not trusted.
- `{message: "Invalid data", data: "Reason", code: 9}` – invalid/incomplete request data.

#### signAndPublishTransaction

This is similar to "`signTransaction"`, but it also broadcasts a transaction to the blockchain.

Example:

```js
const txData = {
  type: 4,
  data: {
    amount: {
      assetId: 'TAC',
      tokens: '1.567',
    },
    fee: {
      assetId: 'TAC',
      tokens: '0.001',
    },
    recipient: 'test',
  },
};

Norsu.signAndPublishTransaction(txData)
  .then(data => {
    //data - a line ready for sending to Tac network's node (server)
  })
  .catch(error => {
    //processing errors
  });
```

REPLY

A reply from Tac network returns as a line containing the entire past transaction.

ERRORS

- Same as `signTransaction`
- `{message: "Filed request", data: "Error description", code: 15}` – a request was signed but not broadcast

#### signTransactionPackage

A package transaction signature. Sometimes several transactions need to be simultaneously signed, and for users' convenience, up to seven transactions at ones could be signed. Only certain types of transactions are permitted:

- `3` – issues a token
- `4` – transfer a token
- `5` – reissues a token
- `6` – burns a token
- `8` – leases TAC
- `9` – cancels the leasing
- `10` – creating an alias for an address in Tac network
- `11` – mass transfer
- `12` - adds, modifies and deletes data entries in the sender's account data storage
- `13` - assigns the dApp script or account script to the sender's account
- `14` - configures sponsorship
- `15` - modifies the asset script
- `16` - calls the function from a dApp script

Example:

```js
const name = 'For Test';
const tx = [
  {
    type: 4,
    data: {
      amount: {
        assetId: 'TAC',
        tokens: '1.567',
      },
      fee: {
        assetId: 'TAC',
        tokens: '0.001',
      },
      recipient: 'test',
    },
  },
  {
    type: 4,
    data: {
      amount: {
        assetId: 'TAC',
        tokens: '0.51',
      },
      fee: {
        assetId: 'TAC',
        tokens: '0.001',
      },
      recipient: 'merry',
    },
  },
];

Norsu.signTransactionPackage(tx, name);
```

Sign two transaction:

- Transfer 1.567 TAC to the alias test
- Transfer 0.1 TAC to the alias merry

REPLY

A unit of two lines – transactions that are signed and ready to broadcast.

ERRORS

Same as in `signTransaction`

##### [Transactions](https://docs.tac.tech/en/building-apps/tac-api-and-sdk/client-libraries/tac-transactions)\*\*

Every user of Tac' network has a state (balances, assets, data, scripts), and every past transactions changes these data. \
In Norsu API it is different from [NODE REST API](https://docs.tac.tech/en/tac-node/node-api/).

`signTransaction`, `signAndPublishTransaction` accept transactions as follows

```js
{
    type: number, //transaction type
    data: {
        ... //transaction data
    }
}
```

Legend keys

- - optional field, data are automatically supplied from Norsu. \
    [x,y] – length limit from x to y. \
    [,x] – length limit to x. \
    [y,] – length limit from y. \
    [x-y] – number from x to y. x/y - x or y. (JLM) - JAVA LONG MAX = 9 223 372 036 854 775 807 \
    MoneyLike - price

MoneyLike could look as:

- `{ tokens: 1, assetId: "TAC" }`
- `{ coins: 100000000, assetId: "TAC" }`;

In both messages, the same price of 1 TAC is indicated. You can easily convert `coins`into `tokens`and back, if you know in what asset the price is indicated, and you have received its precision `tokens = coins / (10 ** precision)` \
If the field contains other types than MoneyLike, for instance, string/MoneyLike , the sum is indicated as a number in  `coins`.

---

###### [Type 3 ISSUE – token issue](https://docs.tac.tech/en/blockchain/transaction-type/issue-transaction)\*\*

- `name` [4, 16] string – token name,
- `description` [0, 1000] string – token description,
- `quantity` [0 - (JLM)] number/string - quantity,
- `precision` [0 - 8] number - precision,
- `reissuable` true|false – re-issuable,
- `*fee` MoneyLike -fee
- `*script` string – [smart asset](https://docs.tac.tech/en/building-apps/smart-contracts/what-is-smart-asset)
- `*senderPublicKey` string – sender's public key in base58
- `*timestamp` number/string – time in ms

Example:

```js
Norsu.signAndPublishTransaction({
  type: 3,
  data: {
    name: 'Best Token',
    description: 'Greate token',
    quantity: 1000000,
    precision: 2,
    reissuable: true,
    fee: {
      tokens: '1',
      assetId: 'TAC',
    },
  },
})
  .then(tx => {
    console.log("Hurray! I've created my asset!!!");
  })
  .catch(error => {
    console.error('Something went wrong', error);
  });
```

In case of a success, we are issuing a new asset in the quantity of 1,000,000, and your balance will show 10,000.00 Best Token

###### [Type 4 TRANSFER – asset transfer](https://docs.tac.tech/en/blockchain/transaction-type/transfer-transaction)\*\*

- `amount` MoneyLike - amount,
- `recipient` string – recipient's address or alias
- `attachment`[,140 bytes] string or byte Array – additional info in text
- `*fee` MoneyLike - fee
- `*senderPublicKey` string – sender's public key in base58
- `*timestamp` number/string – time in ms

Example:

```js
Norsu.signAndPublishTransaction({
  type: 4,
  data: {
    amount: { tokens: '3.3333333', assetId: 'TAC' },
    fee: { tokens: '0.001', assetId: 'TAC' },
    recipient: 'merry',
  },
})
  .then(tx => {
    console.log("Hurray! I've been able to send TAC!!!");
  })
  .catch(error => {
    console.error('Something went wrong', error);
  });
```

###### [Type 5 REISSUE – token reissue](https://docs.tac.tech/en/blockchain/transaction-type/reissue-transaction)\*\*

- `assetId` string - "asset ID",
- `quantity` [0 - (JLM)] number/string/MoneyLike - quantity,
- `reissuable` false – deny reissue
- `*fee` MoneyLike -fee
- `*senderPublicKey` string – sender's public key in base58
- `*timestamp` number/string – time in ms

Example:

```js
Norsu.signAndPublishTransaction({
  type: 5,
  data: {
    quantity: 1000,
    assetId: '8LQW8f7P5d5PZM7GtZEBgaqRPGSzS3DfPuiXrURJ4AJS',
    reissuable: true,
    fee: {
      tokens: '1',
      assetId: 'TAC',
    },
  },
})
  .then(tx => {
    console.log("Hurray! I've reissued my asset!!!");
  })
  .catch(error => {
    console.error('Something went wrong', error);
  });
```

In case of a success, we are re-issuing a new asset in the quantity of 1,000,000, and your balance will show 10,000.00 Best Token

###### [Type 6 BURN – burning tokens](https://docs.tac.tech/en/blockchain/transaction-type/burn-transaction)\*\*

- `assetId` string – asset ID,
- `amount` [0 - (JLM)] number/string/MoneyLike - quantity,
- `*fee` MoneyLike -fee
- `*senderPublicKey` string - sender's public key in base58
- `*timestamp` number/string – time in ms

Example:

```js
Norsu.signAndPublishTransaction({
  type: 6,
  data: {
    amount: 1000,
    assetId: '8LQW8f7P5d5PZM7GtZEBgaqRPGSzS3DfPuiXrURJ4AJS',
    fee: {
      tokens: '0.001',
      assetId: 'TAC',
    },
  },
})
  .then(tx => {
    console.log("Hurray! I've burned unneeded tokens!!!");
  })
  .catch(error => {
    console.error('Something went wrong', error);
  });
```

In case of a success, 1,000 coins `are burned`.

###### [Type 8 LEASE - Leasing](https://docs.tac.tech/en/blockchain/transaction-type/lease-transaction)\*\*

- `recipient` string – recipient's address or alias,
- `amount` [0 - (JLM)] number/string/MoneyLike - quantity,
- `*fee` MoneyLike -fee
- `*senderPublicKey` string - sender's public key in base58
- `*timestamp` number/string – time in ms

Example:

```js
Norsu.signAndPublishTransaction({
  type: 8,
  data: {
    amount: 1000,
    recipient: 'merry',
    fee: {
      tokens: '0.001',
      assetId: 'TAC',
    },
  },
})
  .then(tx => {
    console.log("Hurray! I've been able to lease tokens!!!");
  })
  .catch(error => {
    console.error('Something went wrong', error);
  });
```

In case of a success, 0.00001000 TAC is leased.

###### [Type 9 LEASE CANCEL – cancel leasing](https://docs.tac.tech/en/blockchain/transaction-type/lease-cancel-transaction)\*\*

- `leaseId` string – leasing transaction ID,
- `*fee` MoneyLike -fee
- `*senderPublicKey` string - sender's public key in base58
- `*timestamp` number/string – time in ms

Example:

```js
Norsu.signAndPublishTransaction({
  type: 9,
  data: {
    leaseId: '6frvwF8uicAfyEfTfyC2sXqBJH7V5C8he5K4YH3BkNiS',
    fee: {
      tokens: '0.001',
      assetId: 'TAC',
    },
  },
})
  .then(tx => {
    console.log("Hurray! I've cancelled leasing!!!");
  })
  .catch(error => {
    console.error('Something went wrong ', error);
  });
```

In case of a success, the leasing transaction is cancelled.

###### [Type 10 CREATE ALIAS – creating an alias for an address](https://docs.tac.tech/en/blockchain/transaction-type/create-alias-transaction)\*\*

- `alias`[4, 30] string - alias
- `*fee` MoneyLike -fee
- `*senderPublicKey` string - sender's public key in base58
- `*timestamp` number/string – time in ms

Example:

```js
Norsu.signAndPublishTransaction({
  type: 10,
  data: {
    alias: 'testAlias',
    fee: {
      tokens: '0.001',
      assetId: 'TAC',
    },
  },
})
  .then(tx => {
    console.log('Hurray! Now I have an alias!!!');
  })
  .catch(error => {
    console.error('Something went wrong', error);
  });
```

In case of a success, an alias (another name) is created.

###### [Type 11 MASS TRANSFER - a mass transfer of an asset](https://docs.tac.tech/en/blockchain/transaction-type/mass-transfer-transaction)\*\*

- `totalAmount` moneyLike – total to be sent // instead of calculating the amount you may insert { assetId: "ID of the asset to be sent", coins: 0},
- `transfers`  a mass of objects
  { `recipient`: string - address/alias, amount: number/string/moneyLike }
- `*fee` MoneyLike -fee
- `attachment` [,140 bytes в base58] string – additional info
- `*senderPublicKey` string - sender's public key in base58
- `*timestamp` number/string – time in ms

Example:

```js
Norsu.signAndPublishTransaction({
  type: 11,
  data: {
    totalAmount: { assetId: 'TAC', coins: 0 },
    transfers: [
      { recipient: 'alias1', amount: '200000' },
      { recipient: 'alias2', amount: '200000' },
    ],
    fee: {
      tokens: '0.002',
      assetId: 'TAC',
    },
  },
})
  .then(tx => {
    console.log("Hurray! I've sent hi to my friends!!!");
  })
  .catch(error => {
    console.error('Something went wrong', error);
  });
```

In case of a success, 0.002 TAC will be sent to alias1 and alias2.

###### [Type 12 DATA TRANSACTION - saving data](https://docs.tac.tech/en/blockchain/transaction-type/data-transaction)\*\*

- `data`  mass of objects
  - `type` "binary"/string/"integer"/"boolean" - type,
  - `key` string – field name
  - `value` /string/string/number/boolean depends on the type
- `*fee` MoneyLike - fee
- `*senderPublicKey` string - sender's public key in base58
- `*timestamp` number/string – time in ms

Field:

```js
Norsu.signAndPublishTransaction({
  type: 12,
  data: {
    data: [
      {
        key: 'string',
        value: 'testVdfgdgf dfgdfgdfg dfg dfg al',
        type: 'string',
      },
      {
        key: 'binary',
        value: 'base64:AbCdAbCdAbCdAbCdAbCdAbCdAbCdAbCdAbCdAbCdAbCd',
        type: 'binary',
      },
      { key: 'integer', value: 20, type: 'integer' },
      { key: 'boolean', value: false, type: 'boolean' },
    ],
    fee: {
      tokens: '0.01',
      assetId: 'TAC',
    },
  },
})
  .then(tx => {
    console.log("Hurray! I've saved data!!!");
  })
  .catch(error => {
    console.error('Something went wrong', error);
  });
```

In case of a success, new data will be stored in the state.

###### [Type 13 SET SCRIPT – scripting an account](https://docs.tac.tech/en/blockchain/transaction-type/set-script-transaction)\*\*

- `script` string - [script](https://docs.tac.tech/en/building-apps/smart-contracts/tac-smart-contracts-overview)
- `*fee` MoneyLike -fee
- `*senderPublicKey` string - sender's public key in base58
- `*timestamp` number/string – time in ms

For cancelling a script the field `script` has to be `null`. [Script development on RIDE](https://tac-ide.com/)

Example:

```js
Norsu.signAndPublishTransaction({
  type: 13,
  data: {
    script: '',
    fee: {
      tokens: '0.04',
      assetId: 'TAC',
    },
  },
})
  .then(tx => {
    console.log("Hurray! I've cancelled a script!!!");
  })
  .catch(error => {
    console.error('Something went wrong', error);
  });
```

In case of a success, the script will be removed from the account.

Example 2:

```js
Norsu.signAndPublishTransaction({
  type: 13,
  data: {
    script:
      'base64:AQa3b8tHAQa3b8tHAQa3b8tHAQa3b8tHAQa3b8tHAQa3b8tHAQa3b8tHAQa3b8tHAQa3b8tHAQa3b8tHAQa3b8tH',
    fee: {
      tokens: '0.01',
      assetId: 'TAC',
    },
  },
})
  .then(tx => {
    console.log("Hurray! I've added a script!!!");
  })
  .catch(error => {
    console.error('Something went wrong', error);
  });
```

In case of a success, a new script will be added to the account, allowing any transactions without a signature (be careful!).

###### [Type 14 Sponsored Fee Transaction - Sponsorship](https://docs.tac.tech/en/blockchain/transaction-type/sponsor-fee-transaction)\*\*

- `minSponsoredAssetFee` MoneyLike – fee price in the asset.
- `*fee` MoneyLike - fee
- `*senderPublicKey` string - sender's public key in base58
- `*timestamp` number/string – time in ms

Example:

```js
Norsu.signAndPublishTransaction({
  type: 14,
  data: {
    minSponsoredAssetFee: {
      assetId: '6frvwF8uicAfyEfTfyC2sXqBJH7V5C8he5K4YH3BkNiS',
      tokens: 0.1,
    },
    fee: {
      tokens: '1',
      assetId: 'TAC',
    },
  },
})
  .then(tx => {
    console.log("Hurray! I've become a sponsor!!!");
  })
  .catch(error => {
    console.error('Something went wrong', error);
  });
```

In case of a success, a transfer fee can be paid in the asset.

###### [Type 15 SET ASSET SCRIPT – setting a script to an asset](https://docs.tac.tech/en/blockchain/transaction-type/set-asset-script-transaction)\*\*

- `assetId` string – asset ID
- `script` string – [script](https://docs.tac.tech/en/building-apps/smart-contracts/what-is-smart-asset)
- `*fee` MoneyLike – fee
- `*senderPublicKey` string - sender's public key in base58
- `*timestamp` number/string – time in ms

It's now impossible to cancel a script, you can only add a new script. [Script development on RIDE](https://tac-ide.com/)

Example:

```js
Norsu.signAndPublishTransaction({
  type: 15,
  data: {
    assetId: '',
    script: 'base64:AQa3b8tH',
    fee: {
      tokens: '0.01',
      assetId: 'TAC',
    },
  },
})
  .then(tx => {
    console.log('Hurray! I have reset a script to the asset!!!');
  })
  .catch(error => {
    console.error('Something went wrong', error);
  });
```

In case of a success, the asset's script will be reset.

###### [Type 16 SCRIPT INVOCATION - call account script function](https://docs.tac.tech/en/blockchain/transaction-type/invoke-script-transaction)\*\*

- `dApp` string – address script account
- `call` object –
  - `function` string function name
  - `args` array
    - `type` "binary"/string/"integer"/"boolean" - type,
    - `value` /string/string/number/boolean - value for type
- `*fee` MoneyLike – fee
- `*payment` array MoneyLike
- `*senderPublicKey` string - public key in base58
- `*timestamp` number/string - number/string – time in ms

Example:

```js
Norsu.signAndPublishTransaction({
  type: 16,
  data: {
    fee: {
      tokens: '0.05',
      assetId: 'TAC',
    },
    dApp: '3N27HUMt4ddx2X7foQwZRmpFzg5PSzLrUgU',
    call: {
      function: 'tellme',
      args: [
        {
          type: 'string',
          value: 'Will?',
        },
      ],
    },
    payment: [{ assetId: 'TAC', tokens: 2 }],
  },
})
  .then(tx => {
    console.log('Ура! Я выполнил скрипт!!!');
  })
  .catch(error => {
    console.error('Что-то пошло не так', error);
  });
```

In case of a success, invoke script function `tellme` in testnet account `3N27HUMt4ddx2X7foQwZRmpFzg5PSzLrUgU`

##### [Calculating transaction fees](https://docs.tac.tech/en/blockchain/transaction/transaction-fee)\*\*

---

#### signOrder\*\*

Norsu's method for signing an order to the matcher. As input, it accepts an object similar to a transaction like this:

```js
{
    type: 1002,
    data: {
        ...data
    }
}
```

- `*version` 1,2,3
- `amount` MoneyLike - amount
- `price` MoneyLike - price
- `orderType` 'sell'/'buy' – order type
- `matcherFee` MoneyLike - fee (0.003 TAC minimum),
- `matcherPublicKey` string - the public key of the exchange service
- `expiration` string/number – the order's expiration time
- `*timestamp` string/number - current time
- `*senderPublicKey` string - public key in base58

Example:

```js
Norsu.signOrder({
  type: 1002,
  data: {
    matcherPublicKey: '7kPFrHDiGw1rCm7LPszuECwWYL3dMf6iMifLRDJQZMzy',
    orderType: 'sell',
    expiration: Date.now() + 100000,
    amount: {
      tokens: '100',
      assetId: 'TAC',
    },
    price: {
      tokens: '0.01',
      assetId: '8LQW8f7P5d5PZM7GtZEBgaqRPGSzS3DfPuiXrURJ4AJS',
    },
    matcherFee: {
      tokens: '0.03',
      assetId: 'TAC',
    },
  },
})
  .then(tx => {
    console.log("Hurray! I've signed an order!!!");
  })
  .catch(error => {
    console.error('Something went wrong', error);
  });
```

REPLY: A line with data for sending to the matcher.

ERRORS:

- `{ message: "User denied message", code: 10 }` – the user rejected the request
- `{ message: "Api rejected by user", code: 12 }` – The website is not trusted
- `{ message: "Invalid data", data: "Reason", code: 9 }` - invalid/incomplete request data

#### signAndPublishOrder

Norsu's method for creating an order to the matcher is identical to `signOrder`, but it also tries to send data to the matcher.

REPLY: the matcher's reply line about successful creation of an order.

ERRORS:

- Same as for `signOrder`
- `{message: "Filed request", data: "Error description", code: 15}` – a request has been signed, but not sent to the matcher

#### signCancelOrder

Norsu's method for signing cancellation of an order to the matcher. As input, it accepts an object similar to a transaction like this:

```js
{
    type: 1003,
    data: {
        ...data
    }
}
```

- `id` string – order ID
- `*senderPublicKey` - string - sender's public key in base58

Example:

```js
Norsu.signCancelOrder({
  type: 1003,
  data: {
    id: '31EeVpTAronk95TjCHdyaveDukde4nDr9BfFpvhZ3Sap',
  },
});
```

REPLY: A data line for sending to the matcher.

ERRORS:

- `{ message: "User denied message", code: 10 }` – the user rejected the request

changeLog.md
## Tac Keeper

#### version 1.1.12

- Fix links to tac.exchange

#### version 1.1.11

- Fix debounce on delete account btn
- Add backup accounts store

#### version 1.1.10

- Added methods to sign and verify custom data
- Added Stagenet in network selector
- simplified the creation of the new accounts
- accounts are now sorted in the list by last activity
- added Privacy Policy confirmation on the registration screen
- fixed bug of empty InvokeScript function error
- fixed network byte loss , while switching between different custom networks

#### version 1.1.9

- Norsu proxy
- Update fixed libs

#### version 1.1.8

- Add auto calculate fee
- Fix auth api
- Orders UI
- Update major libs versions with keeper refactor (updated crypto protocol)

#### version 1.1.7

- Order v3 support
- Fix auto payment
- Add script invocation in package
- Fix some texts

#### version 1.1.6

- Add encrypt and decrypt API by DH
- Fix some UI problems

#### version 1.1.5

- Add notifications
- Update development invocation script
- Fix publicState eventschrome
- Fix build with no lokalize api key
- Some UI fixes
- Messages refactoring

#### version 1.1.4

- Idle configurable
- Some ui fixes
- Tokenomika auto orders

#### version 1.1.3

- Fix bug with attachments from tac lib

#### version 1.1.2

- Add script invocation type: 16
- Add initial promise
- Update documentation

#### version 1.1.1

- Remove Google Analytics
- Update package,json meta data
- Fix delete wallet
- Fix auth protocol host

#### version 1.1.0

- Add custom node for developer
- Add auto sign transactions, with limits by time and tokens
- Some visual bug fixes

#### version 1.0.9

- Fix memory leaks
- Attachment as byte Array in tx type 4 and type 11

#### version 1.0.8

- Fix init badge on restart keeper
- Refactor requests and change design
- Init cancel lease request (fetch transaction lease)
- Fix public state permissions on empty keeper
- Add list of waiting requests
- Add batch of txs
- Extends remote config
- Add autoClick disable setting
- Add ru/en documentation
- Refactor errors
- Opera build scripts update.

---

#### version 1.0.7

- Expiration requests
- Add black/white lists
- Add user permissions
- Add edit permissions page
- Add smart asset transactions
- Some design fixes
- Refactor current active message
- Add to keeper public state tx versions
- Update signature lib
- Update dictionary JSON
- Update api massTransfer tx format
- Fix auth api


src/constants.js
export const TACKEEPER_DEBUG = process.env.NODE_ENV !== 'production';
export const TACKEEPER_ENV = process.env.NODE_ENV || 'development';

export const CONFIG_URL =
  'https://raw.githubusercontent.com/tactokens/norsu/main/json/tac_norsu_blacklist.json';

export const allowMatcher = ['dex.tokenomica.com', 'vfa.tokenomica.com'];

export const MSG_STATUSES = {
  UNAPPROVED: 'unapproved',
  SIGNED: 'signed',
  PUBLISHED: 'published',
  FAILED: 'failed',
  REJECTED: 'rejected',
  REJECTED_FOREVER: 'rejected_forever',
  SHOWED_NOTIFICATION: 'showed_notify',
  NEW_NOTIFICATION: 'new_notify',
};

export const STATUS = {
  ERROR: -1,
  OK: 1,
  PENDING: 0,
  UPDATED: 2,
};

export const DEFAULT_CONFIG = {
  CONFIG: {
    update_ms: 30000,
  },
  NETWORKS: ['mainnet', 'testnet', 'stagenet', 'custom'],
  NETWORK_CONFIG: {
    testnet: {
      code: '$',
      // server: 'https://testnet.tactoken.io/',
      server: 'https://testnet.tactoken.io/',
      matcher: 'https://testnet-explorer.tactoken.io/',
    },
    mainnet: {
      code: '%',
      server: 'https://nodes.tactoken.io/',
      matcher: 'https://explorer.tactoken.io/',
    },
    stagenet: {
      code: 'S',
      server: 'https://stagenet.tactoken.io/',
      matcher: 'https://stagenet-explorer.tactoken.io/',
    },
    custom: {
      code: '',
      server: '',
      matcher: '',
    },
  },
  MESSAGES_CONFIG: {
    message_expiration_ms: 30 * 60 * 1000,
    update_messages_ms: 30 * 1000,
    max_messages: 100,
    notification_title_max: 20,
    notification_interval_min: 30 * 1000,
    notification_message_max: 250,
  },
  PACK_CONFIG: {
    max: 7,
    allow_tx: [3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16],
  },
  IDLE: {
    idle: 0,
    '5m': 5 * 60 * 1000,
    '10m': 10 * 60 * 1000,
    '20m': 20 * 60 * 1000,
    '40m': 40 * 60 * 1000,
    '1h': 60 * 60 * 1000,
  },
};

export const DEFAULT_FEE_CONFIG_URL =
  'https://raw.githubusercontent.com/tactokens/norsu/main/json/fee.json';

export const DEFAULT_FEE_CONFIG = {
  smart_asset_extra_fee: 400000,
  smart_account_extra_fee: 400000,
  calculate_fee_rules: {
    default: {
      add_smart_asset_fee: true,
      add_smart_account_fee: true,
      min_price_step: 100000,
      fee: 100000,
    },
    3: {
      fee: 100000000,
      nftFee: 100000,
    },
    7: {
      add_smart_account_fee: false,
      fee: 300000,
    },
    11: {
      price_per_transfer: 50000,
    },
    12: {
      price_per_kb: 100000,
    },
    13: {
      fee: 1000000,
    },
    15: {
      fee: 100000000,
    },
    16: {
      fee: 500000,
    },
    17: {
      fee: 100000,
    },
  },
};
